### PR TITLE
FortiOS System Fimrware Upgrade Module

### PR DESCRIPTION
**Pull request by fortinetps**
##### SUMMARY
Fortinet is adding Ansible support for FortiOS and FortiGate products. This module follows the same structure, guidelines and ideas given in previous approved module for a parallel feature of FortiGate (user device): #56447
In this case we are providing a different functionality: "Fortios System Firmware Upgrade".

Please note that this will be part of other modules to come for FortiGate, including different functionalities: system, wireless-controller, firewall, webfilter, ips, web-proxy, wanopt, application, dlp spamfilter, log, vpn, certificate, user, dnsfilter, antivirus, report, waf, authentication, switch controller, endpoint-control and router. We plan to follow the same style, structure and usage as in the previous module in order to make it easier to comply with Ansible guidelines.

This PR has been updated according to suggestions and code reviews given in previous submissions of Fortinet modules.

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
fortios_system_firmware_upgrade

##### ANSIBLE VERSION
ansible 2.9.0.dev0
python version = 3.7.4 (default, Jul 13 2019, 14:04:11) [GCC 8.3.0]

### CODE REVIEW

**Comment by fortinetps** [comment]
Hi @NilashishC, I am working with Miguel, I have this fortios_system_firmware_upgrade module, which is also created by Miguel's generator, and use FortiOS API to upgrade the FortiOS/FortiGate firmware. Let me know if you have any questions or comments. Thank you very much!

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    username:
+        description:
+            - FortiOS or FortiGate username.
+        type: str
+        required: true
+    password:
+        description:
+            - FortiOS or FortiGate password.
+        type: str
+        default: ""
```
Comment (NilashishC) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:58]: Why is the `default: ""` needed for this key?

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        default: root
+    https:
+        description:
+            - Indicates if the requests towards FortiGate must use HTTPS protocol.
+        type: bool
+        default: true
+    system_firmware:
+        description:
+            - Perform firmware upgrade with local firmware file.
+        default: null
```
Comment (NilashishC) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:84]: Same as above.

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 1 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:58]: In legacy mode (local connection), without using httpapi plugin, password is needed for login to the device to perform API call.

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 2 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:84]: system_firmware is user input data so I set it default to null.

**Review by fortinetps — COMMENTED** [review]
The module itself, including the unit test, require fortiosapi python package to run or perform unit test. Unit test module failed to pass sanity test since fortiosapi python package is not included, could you please help?

**Comment by NilashishC** [comment]
@fortinetps could you have a look at the unit tests in any of the PRs raised by @mamunozgonzalez. For example - https://github.com/ansible/ansible/pull/61037

**Comment by fortinetps** [comment]
@NilashishC thank you for your feedback, I will check today!

**Comment by fortinetps** [comment]
Unit test requires fortiosapi
I appended the following line at the end of "/test/runner/requirements/units.txt"
fortiosapi==0.11.1
it passes the ansible-unit test.
Thanks!

**Comment by NilashishC** [comment]
@fortinetps CI is failing. Could you please check?

**Comment by NilashishC** [comment]
@fortinetps I have merged the dependent PR and re-kicked the tests.

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+        type: dict
+        suboptions:
+            file_content:
+                description:
+                    - "Provided when uploading a file: base64 encoded file data. Must not contain whitespace or other invalid base64 characters. Must be
+                       included in HTTP body."
+                type: str
+            filename:
```
Comment (rohitthakur2590) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:94]: required is missing

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    dictionary = {}
+
+    for attribute in option_list:
+        if attribute in json and json[attribute] is not None:
+            dictionary[attribute] = json[attribute]
+
+    return dictionary
+
+
+def underscore_to_hyphen(data):
```
Comment (rohitthakur2590) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py]: why we even need this ?

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+        description:
+            - Indicates if the requests towards FortiGate must use HTTPS protocol.
+        type: bool
+        default: true
+    ssl_verify:
+        description:
+            - Ensures FortiGate certificate must be verified by a proper CA.
+        type: bool
+        default: true
+    system_firmware:
```
Comment (rohitthakur2590) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:79]: required is missing , as its true in argspec

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+module: fortios_system_firmware_upgrade
+short_description: Perform firmware upgrade with local firmware file in Fortinet's FortiOS and FortiGate.
+description:
+    - This module is able to configure a FortiGate or FortiOS (FOS) device by allowing the
+      user to set and modify system feature and firmware category.
+      Examples include all parameters and values need to be adjusted to datasources before usage.
+      Tested with FOS v6.0.2
+version_added: "2.9"
+author:
+    - Miguel Angel Munoz (@mamunozgonzalez)
```
Comment (rohitthakur2590) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:36]: is the module author correct ?

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'metadata_version': '1.1'}
+
+DOCUMENTATION = '''
+---
+module: fortios_system_firmware_upgrade
+short_description: Perform firmware upgrade with local firmware file in Fortinet's FortiOS and FortiGate.
+description:
```
Comment (rohitthakur2590) [inline lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:28]: this seems to be inaccurate based on what the module does

Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:28]: updated, please check, thank you!

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 6 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:36]: updated, add my name, please check

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 5 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:79]: updated, please check

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 4 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py]: removed, please check

File: lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py (hunk 3 above)
Comment (fortinetps) [reply lib/ansible/modules/network/fortios/fortios_system_firmware_upgrade.py:94]: updated, please check

**Review by rohitthakur2590 — APPROVED** [review]

**Pull request merged**